# Used Car Price Prediction


## Objective

The objective of this competition is to **predict the price of used cars** based on various attributes, such as brand, model, manufacturing year, mileage, <br>
fuel type, engine specifications, and more. Participants are required to build a machine learning model that accurately estimates car prices using the <br>
provided dataset, which contains information about each vehicle's features and characteristics. The model's performance will be evaluated based on <br>
the **Root Mean Squared Error (RMSE)** between the predicted prices and the actual prices.

## Dataset

- **Brand & Model**: Identify the brand or company name along with the specific model of each vehicle.
- **Model Year**: Discover the manufacturing year of the vehicles, crucial for assessing depreciation and technology advancements.
- **Mileage**: Obtain the mileage of each vehicle, a key indicator of wear and tear and potential maintenance requirements.
- **Fuel Type**: Learn about the type of fuel the vehicles run on, whether it's gasoline, diesel, electric, or hybrid.
- **Engine Type**: Understand the engine specifications, shedding light on performance and efficiency.
- **Transmission**: Determine the transmission type, whether automatic, manual, or another variant.
- **Exterior & Interior Colors**: Explore the aesthetic aspects of the vehicles, including exterior and interior color options.
- **Accident History**: Discover whether a vehicle has a prior history of accidents or damage, crucial for informed decision-making.
- **Clean Title**: Evaluate the availability of a clean title, which can impact the vehicle's resale value and legal status.
- **Price**: Access the listed prices for each vehicle, aiding in price comparison and budgeting.

## Initial Data Exploration

### What the data looks like

In [55]:
# Load necessary libraries
import pandas as pd
import numpy as np
import warnings
warnings.filterwarnings("ignore")

In [56]:
# Load the dataset
train_df = pd.read_csv('/Users/shudizhao/data_projects/used_car_price_prediction/train.csv')
test_df = pd.read_csv('/Users/shudizhao/data_projects/used_car_price_prediction/test.csv')

# Preview the data
print("Train Data:")
train_df.head()

Train Data:


id          brand              model  model_year  milage      fuel_type  \
0   0           MINI      Cooper S Base        2007  213000       Gasoline   
1   1        Lincoln              LS V8        2002  143250       Gasoline   
2   2      Chevrolet  Silverado 2500 LT        2002  136731  E85 Flex Fuel   
3   3        Genesis   G90 5.0 Ultimate        2017   19500       Gasoline   
4   4  Mercedes-Benz        Metris Base        2021    7388       Gasoline   

                                              engine  \
0       172.0HP 1.6L 4 Cylinder Engine Gasoline Fuel   
1       252.0HP 3.9L 8 Cylinder Engine Gasoline Fuel   
2  320.0HP 5.3L 8 Cylinder Engine Flex Fuel Capab...   
3       420.0HP 5.0L 8 Cylinder Engine Gasoline Fuel   
4       208.0HP 2.0L 4 Cylinder Engine Gasoline Fuel   

                     transmission ext_col int_col  \
0                             A/T  Yellow    Gray   
1                             A/T  Silver   Beige   
2                             A/T    Blue    Gray   
3  Transmission w/Dual Shift Mode   Black   Black   
4                     7-Speed A/T   Black   Beige   

                                 accident clean_title  price  
0                           None reported         Yes   4200  
1  At least 1 accident or damage reported         Yes   4999  
2                           None reported         Yes  13900  
3                           None reported         Yes  45000  
4                           None reported         Yes  97500

In [57]:
print("\nTest Data:")
test_df.head()


Test Data:


id brand                 model  model_year  milage fuel_type  \
0  188533  Land        Rover LR2 Base        2015   98000  Gasoline   
1  188534  Land     Rover Defender SE        2020    9142    Hybrid   
2  188535  Ford    Expedition Limited        2022   28121  Gasoline   
3  188536  Audi         A6 2.0T Sport        2016   61258  Gasoline   
4  188537  Audi  A6 2.0T Premium Plus        2018   59000  Gasoline   

                                              engine        transmission  \
0       240.0HP 2.0L 4 Cylinder Engine Gasoline Fuel         6-Speed A/T   
1  395.0HP 3.0L Straight 6 Cylinder Engine Gasoli...         8-Speed A/T   
2                    3.5L V6 24V PDI DOHC Twin Turbo  10-Speed Automatic   
3                                     2.0 Liter TFSI           Automatic   
4       252.0HP 2.0L 4 Cylinder Engine Gasoline Fuel                 A/T   

           ext_col int_col       accident clean_title  
0            White   Beige  None reported         Yes  
1           Silver   Black  None reported         Yes  
2            White   Ebony  None reported         NaN  
3  Silician Yellow   Black  None reported         NaN  
4             Gray   Black  None reported         Yes

In [58]:
# Check the structure and data types
print("\nTrain Data Info:")
train_df.info()


Train Data Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 188533 entries, 0 to 188532
Data columns (total 13 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   id            188533 non-null  int64 
 1   brand         188533 non-null  object
 2   model         188533 non-null  object
 3   model_year    188533 non-null  int64 
 4   milage        188533 non-null  int64 
 5   fuel_type     183450 non-null  object
 6   engine        188533 non-null  object
 7   transmission  188533 non-null  object
 8   ext_col       188533 non-null  object
 9   int_col       188533 non-null  object
 10  accident      186081 non-null  object
 11  clean_title   167114 non-null  object
 12  price         188533 non-null  int64 
dtypes: int64(4), object(9)
memory usage: 18.7+ MB


### What are the values we have for each columns

* Values Counts for each Categorical Columns
* Distribution for Numerical Columns
* Percentage of missing values in each column

In [59]:
# List of categorical features
categorical_features = ['brand', 'model', 'model_year', 'fuel_type', 'engine', 'transmission', 'ext_col', 'int_col', 'accident', 'clean_title']

# Value counts for each categorical feature
for feature in categorical_features:
    print(f"\nValue counts for {feature}:")
    print(train_df[feature].value_counts())


Value counts for brand:
brand
Ford             23088
Mercedes-Benz    19172
BMW              17028
Chevrolet        16335
Audi             10887
Porsche          10612
Land              9525
Toyota            8850
Lexus             8643
Jeep              6474
Cadillac          4674
RAM               4249
Nissan            3930
Tesla             3738
INFINITI          3276
GMC               3215
Dodge             3133
Mazda             2719
Kia               2497
Lincoln           2423
Subaru            2381
Acura             2282
Honda             2101
Hyundai           2045
Volkswagen        1765
Jaguar            1319
Bentley           1155
MINI              1064
Genesis            969
Buick              940
Maserati           939
Lamborghini        809
Chrysler           727
Volvo              723
Alfa               682
Rivian             590
Rolls-Royce        561
Mitsubishi         551
Pontiac            538
Hummer             520
Ferrari            359
McLaren            243
Ast

In [60]:
numerical_features = ['milage', 'price']

# Descriptive statistics for numerical features
print("\nDescriptive statistics for numerical features:")
print(train_df[numerical_features].describe().round(2))


Descriptive statistics for numerical features:
          milage       price
count  188533.00   188533.00
mean    65705.30    43878.02
std     49798.16    78819.52
min       100.00     2000.00
25%     24115.00    17000.00
50%     57785.00    30825.00
75%     95400.00    49900.00
max    405000.00  2954083.00


In [61]:
def calculate_missing_percentage(train_df): 
    # Calculate the percentage of missing values
    missing_values_percentage = ((train_df.isnull().sum() / len(train_df)) * 100).round(2)

    # Get data types of each column
    data_types = train_df.dtypes

    # Combine the missing values and data types into a DataFrame for easier viewing
    missing_and_dtypes = pd.DataFrame({
        'Missing Percentage': missing_values_percentage,
        'Data Type': data_types
    })

    print("\nMissing Percentage and Data Types for Each Column:")
    print(missing_and_dtypes)
calculate_missing_percentage(train_df)


Missing Percentage and Data Types for Each Column:
              Missing Percentage Data Type
id                          0.00     int64
brand                       0.00    object
model                       0.00    object
model_year                  0.00     int64
milage                      0.00     int64
fuel_type                   2.70    object
engine                      0.00    object
transmission                0.00    object
ext_col                     0.00    object
int_col                     0.00    object
accident                    1.30    object
clean_title                11.36    object
price                       0.00     int64


### Quick Summary

1. clean_title with 11.36% missing values, and no False value. (how to impute? Since we are not sure of null means False, Let's consider it as "Unknown")
2. accident has 1.3% missing values, and fuel_type has 2.7% missing values(Mode imputation is a good option when the missing percentage is low, and the column is likely to follow a common distribution.)
3. Many columns has value "–", let's convert it to Null and fill with Mode as well.
4. 


## Feature Engineering

* Wrap up it into a function so we can simply apply to the test data

In [65]:
# Load the dataset
train_df = pd.read_csv('/Users/shudizhao/data_projects/used_car_price_prediction/train.csv')
test_df = pd.read_csv('/Users/shudizhao/data_projects/used_car_price_prediction/test.csv')

In [66]:
# Count the number of rows with "–" in each column
dash_counts = (train_df == '–').sum()

# Display the count for each column
print("Number of rows with '–' for each column:")
print(dash_counts)

# Replace "–" with NaN (missing values)
train_df.replace('–', np.nan, inplace=True)

# Now, "–" values have been converted to NaN
calculate_missing_percentage(train_df)
# Step 1: Fill missing values in 'clean_title' with "Unknown"
train_df['clean_title'].fillna('Unknown', inplace=True)

for column in train_df.columns:
    if (column != 'clean_title') & (column != 'price'):  # Exclude 'clean_title' since it's already handled
        train_df[column].fillna(train_df[column].mode()[0], inplace=True)

# Verify missing values after filling
print("Missing values after applying the strategy:")
calculate_missing_percentage(train_df)

Number of rows with '–' for each column:
id                 0
brand              0
model              0
model_year         0
milage             0
fuel_type        781
engine           925
transmission      87
ext_col          366
int_col         4527
accident           0
clean_title        0
price              0
dtype: int64

Missing Percentage and Data Types for Each Column:
              Missing Percentage Data Type
id                          0.00     int64
brand                       0.00    object
model                       0.00    object
model_year                  0.00     int64
milage                      0.00     int64
fuel_type                   3.11    object
engine                      0.49    object
transmission                0.05    object
ext_col                     0.19    object
int_col                     2.40    object
accident                    1.30    object
clean_title                11.36    object
price                       0.00     int64
Missing values afte

## Model Selection

In [ ]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.linear_model import LinearRegression, Ridge, Lasso
from sklearn.ensemble import RandomForestRegressor
from xgboost import XGBRegressor
from catboost import CatBoostRegressor
from sklearn.svm import SVR
from sklearn.metrics import mean_squared_error
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import Pipeline

# Define a function to train and evaluate models with preprocessed data and calculate cross-validation RMSE
def train_and_evaluate_model(train_df, test_df, model_name, cv_folds=5):
    
    # 1. Split the data into features (X) and target (y)
    X = train_df.drop(columns=['price'])
    y = train_df['price']

    # 2. Train-test split on the training data
    X_train, X_val, y_train, y_val = train_test_split(X, y, test_size=0.2, random_state=42)

    # 3. Select the model based on input
    if model_name == "LinearRegression":
        model = LinearRegression()
    elif model_name == "Ridge":
        model = Ridge(alpha=1.0)
    elif model_name == "Lasso":
        model = Lasso(alpha=0.1)
    elif model_name == "RandomForest":
        model = RandomForestRegressor(n_estimators=100, random_state=42)
    elif model_name == "XGBoost":
        model = XGBRegressor(n_estimators=100, learning_rate=0.05, random_state=42)
    elif model_name == "CatBoost":
        model = CatBoostRegressor(verbose=0, random_state=42)
    elif model_name == "SVM":
        model = Pipeline([('scaler', StandardScaler()), ('svr', SVR(kernel='rbf'))])
    else:
        raise ValueError("Model not recognized")

    # 4. Train the model on the training set
    model.fit(X_train, y_train)

    # 5. Predict on the validation data and calculate RMSE
    y_val_pred = model.predict(X_val)
    val_rmse = np.sqrt(mean_squared_error(y_val, y_val_pred))

    # 6. Cross-validation to get the cross-validated RMSE
    cv_scores = cross_val_score(model, X, y, cv=cv_folds, scoring='neg_mean_squared_error')
    cv_rmse = np.sqrt(-cv_scores.mean())

    # 7. Predict on the test data (you can also use the test set later for submission)
    test_X = test_df.drop(columns=['price'])
    y_test_pred = model.predict(test_X)
    
    # 8. Return validation RMSE, cross-validation RMSE, and predictions for the test set
    return val_rmse, cv_rmse, y_test_pred

In [ ]:
# Example usage
models_to_test = ["LinearRegression", "Ridge", "Lasso", "RandomForest", "XGBoost", "CatBoost", "SVM"]

# Loop through all models and evaluate
for model_name in models_to_test:
    val_rmse, cv_rmse, y_test_pred = train_and_evaluate_model(train_df, test_df, model_name, cv_folds=5)
    print(f"Model: {model_name} - Validation RMSE: {val_rmse:.4f} - Cross-Validation RMSE: {cv_rmse:.4f}")

## Hyperparameter Tuning

	•	Industry Standard: Select the best-performing models first, then perform hyperparameter tuning on these models.
	•	Why: Tuning is time-consuming, so focusing on the top models ensures efficiency.
	•	Tools: Use Grid Search, Random Search, or Bayesian Optimization for tuning after selecting the best models.

## Ensemble Methods

# Eveluate and Submit